market_maker_keeper/erisx_market_maker_keeper.py

```diff
@@ -18,19 +18,66 @@
 import argparse
 import logging
 import sys
+import threading
+import os
+import signal
+
 import time
-import json
 from decimal import Decimal
 
 from pyexchange.erisx import ErisxApi
 from pyexchange.model import Order
 
 from pymaker.numeric import Wad
+from pymaker.lifecycle import Lifecycle
 
 from market_maker_keeper.cex_api import CEXKeeperAPI
 from market_maker_keeper.order_book import OrderBookManager
 
 
+class ErisXLifecycle(Lifecycle):
+
+    def _start_every_timer(self, idx: int, frequency_in_seconds: int, callback):
+        self.count = 0
+        self._socket_closed = False
+
+        def setup_timer(delay):
+            timer = threading.Timer(delay, func)
+            timer.daemon = True
+
+            self._start_thread_safely(timer)
+
+        def func():
+            try:
+                if not self.terminated_internally and not self.terminated_externally and not self.fatal_termination:
+                    def on_start():
+                        self.logger.debug(f"Processing the timer #{idx}")
+
+                    def on_finish():
+                        self.logger.debug(f"Finished processing the timer #{idx}")
+                        self.count = 0
+
+                    if not callback.trigger(on_start, on_finish):
+                        self.count += 1
+                        if self.count >= 20:
+                            self.logger.debug(f"killing lifecycle as parent is no longer running")
+                            self.terminated_externally = True
+                            self._socket_closed = True
+                        self.logger.debug(f"Ignoring timer #{idx} as previous one is already running")
+                else:
+                    self.logger.debug(f"Ignoring timer #{idx} as keeper is already terminating")
+                    if self._socket_closed:
```

Comment:
This block of code is just here to handle the case where a previous timer iteration hasn't completed yet, which is perfectly normal when the timer callback is invoked every 1 second.

Reply:
Yep, the issue I was running into was that the keeper logic was spawned as a child process from the `lifecycle`, so any kill signals on the client were getting swallowed and we were subsequently trapped with a dead FIX connection, but an infinite timer check loop. That led me to try to figure out a way to make sure that a socket type failure would cause both child processes, and the encompassing lifecycle to both be stopped at the same time.

Reply:
`every` does seem like it would be much cleaner - I am investigating that approach now.

Reply:
I didn't realize there was multi-processing going on.  I'm not comfortable with that, or the way lifecycle handles the asyncio event loop, but those issues are both beyond the scope of this endeavor.